src/index.js

```diff
@@ -0,0 +1,182 @@
+//    // when finished.  This will send a "closed" notification
+//    // to the server.
+//
+//    await conn.close()
+//
+// The `ServerConnection` API operates similarly; the distinction is mainly
+// in which side is expected to send vs receieve during the protocol handshake.
+
+import {
+    assertIsBytes,
+    bytesToHex,
+    hexToBytes,
+    bytesToUtf8,
+    utf8ToBytes
+} from './utils.js';
+
+import { State_UNINITIALIZED, State_ERROR } from './statemachine/index.js';
+import { ClientState_START } from './statemachine/client.js';
+import { ServerState_START } from './statemachine/server.js';
+
+import { OutgoingRecordBuffer, IncomingRecordBuffer } from './recordlayer.js'
+
+// !!!!!!!
+// !!
+// !!   N.B. We have not yet implemented the actual encryption bits!
+// !!
+// !!!!!!!
+//
+// This first version of the code uses the same "framework" as we would use for
+// implementing TLS1.3, including the basic sequence of messages between client and
+// server, and the use of binary encoding in an ArrayBuffer for data processing.
+// Using an implementation with proper crypto should feel identical to using
+// this mock version, except it won't have "Insecure" in the class name...
+
+class InsecureConnection {
+    constructor(psk, pskID, sendCallback) {
+        this.psk = assertIsBytes(psk)
+        this.pskID = assertIsBytes(pskID)
+        this.sendCallback = sendCallback
+        this._state = new State_UNINITIALIZED(this)
+        this._pendingSendData = []
+        this._incomingRecords = new IncomingRecordBuffer()
+        this._outgoingRecords = new OutgoingRecordBuffer()
+        this._lastPromise = Promise.resolve()
+    }
+
+    // Subclasses will override this with some async initialization logic.
+    static async create(psk, pskID, sendCallback) {
+        return new this(psk, pskID, sendCallback)
+    }
+
+    // These are the three public API methods that
+    // consumers can use to connunicate over TLS.
+
+    async send(data) {
+        assertIsBytes(data)
+        await this._serialize(async () => {
+            this._pendingSendData.push(data)
+            await this._state.send()
+        })
+    }
+
+    async recv(data) {
+        assertIsBytes(data)
+        return await this._serialize(async () => {
+            this._incomingRecords.append(data)
+            return await this._state.recv()
+        })
+    }
+
+    async close() {
+        await this._serialize(async () => {
+            await this._state.close()
+        })
+    }
+
+    // Ensure that async functions execute one at a time,
+    // by waiting for the previous call to `_serialize()` to complete
+    // before starting a new one.  This helps ensure that we complete
+    // one state-machine transition before starting to do the next.
+    //
+    // XXX TODO: this risks keeping the result of the last call around
+    // in memory, we should probably "void" it out somehow on completion.
+
+    _serialize(cb) {
```

Comment:
is `_serialize` the right name for this? 🤔  I wonder if I am missing something

Reply:
I mean it in the sense of "one thing after another in strict order" but am open to better names.

Reply:
@vladikoff perhaps `_synchronized` would be a better name, by analogy to Java's `synchronized` keyword?

Reply:
forgot to respond here, yes `synchronized` ++